cosmwasm/enclaves/shared/contract-engine/src/contract_operations.rs

```diff
@@ -223,6 +259,300 @@ fn to_canonical(contract_address: &BaseAddr) -> Result<CanonicalAddr, EnclaveErr
+
+    let mut versioned_env = base_env.into_versioned_env(&engine.get_api_version());
+
+    versioned_env.set_contract_hash(&contract_hash);
+
+    let new_contract_key = generate_contract_key(
+        &canonical_sender_address,
+        &block_height,
+        &contract_hash,
+        &canonical_contract_address,
+        Some(&og_contract_key),
+    )?;
+
+    #[cfg(feature = "random")]
+    set_random_in_env(
+        block_height,
+        &new_contract_key,
+        &mut engine,
+        &mut versioned_env,
+    );
+
+    update_msg_counter(block_height);
+    let result = engine.migrate(&versioned_env, validated_msg);
+
+    *used_gas = engine.gas_used();
+
+    let output = result?;
+
+    engine
+        .flush_cache()
+        .map_err(|_| EnclaveError::FailedFunctionCall)?;
+
+    let output = post_process_output(
+        output,
+        &secret_msg,
+        &canonical_contract_address,
+        versioned_env.get_contract_hash(),
+        reply_params,
+        &canonical_sender_address,
+        false,
+        false,
+    )?;
+
+    // let duration = start.elapsed();
+    // trace!("Time elapsed in encrypt_output: {:?}", duration);
+
+    // todo: can move the key to somewhere in the output message if we want
+
+    let new_contract_key_proof = generate_contract_key_proof(
+        &canonical_contract_address.0 .0,
+        &contract_code.hash(),
+        &og_contract_key,
+        &new_contract_key,
+    );
+
+    debug!(
+        "Migrate success: {:x?}, {:x?}",
+        new_contract_key, new_contract_key_proof
+    );
+
+    Ok(MigrateSuccess {
+        output,
+        new_contract_key,
+        new_contract_key_proof,
+    })
+}
+
+pub fn update_admin(
+    env: &[u8],
+    sig_info: &[u8],
+    current_admin: &[u8],
+    current_admin_proof: &[u8],
+    new_admin: &[u8],
+) -> Result<UpdateAdminSuccess, EnclaveError> {
+    debug!("Starting update_admin");
+
+    let base_env: BaseEnv = extract_base_env(env)?;
+
+    #[cfg(feature = "light-client-validation")]
+    {
+        verify_block_info(&base_env)?;
+    }
+
+    let (sender, contract_address, _block_height, sent_funds) = base_env.get_verification_params();
+
+    let canonical_sender_address = to_canonical(sender)?;
+    let canonical_current_admin_address = CanonicalAddr::from_vec(current_admin.to_vec());
+    let canonical_new_admin_address = CanonicalAddr::from_vec(new_admin.to_vec());
+
+    let canonical_contract_address = to_canonical(contract_address)?;
+
+    if is_hardcoded_contract_admin(
+        &canonical_contract_address,
+        &canonical_current_admin_address,
+        current_admin_proof,
+    ) {
+        debug!(
+            "Found hardcoded admin for update_admin. Cannot update admin for hardcoded contracts."
+        );
+        return Err(EnclaveError::ValidationFailure);
+    }
+
+    let og_contract_key = base_env.get_og_contract_key()?;
+
+    let sender_admin_proof = generate_admin_proof(&canonical_sender_address.0 .0, &og_contract_key);
```

Comment:
Same as with migrate, what prevents an admin of another contract to passing his proof here?

Reply:
Same answer as https://github.com/scrtlabs/SecretNetwork/pull/1449#discussion_r1278528244